packages/kbn-esql-validation-autocomplete/scripts/generate_function_validation_tests.ts

```diff
@@ -430,6 +391,79 @@ function generateEvalCommandTestsForEvalFunction(
       );
     }
   }
+
+  // test that the function can have too many args
+  if (signatures.some(({ minParams }) => minParams)) {
+    // at least one signature is variadic, so no way
+    // to have too many arguments
+    return;
+  }
+
+  // test that additional args are spotted
+
+  const getNumberOfParams = (signature: FunctionDefinition['signatures'][number]) => ({
+    all: signature.params.length,
+    required: signature.params.filter(({ optional }) => !optional).length,
+  });
+
+  // get the signature with the greatest number of params
+  const [first, ...rest] = signatures;
+  let signatureWithGreatestNumberOfParams = first;
+  let { all: maxNumberOfArgs, required: minNumberOfArgs } = getNumberOfParams(first);
+
+  for (const signature of rest) {
+    const numberOfParams = signature.params.length;
+    if (numberOfParams > signatureWithGreatestNumberOfParams.params.length) {
+      signatureWithGreatestNumberOfParams = signature;
+    }
+
+    maxNumberOfArgs = Math.max(maxNumberOfArgs, numberOfParams);
+    const numberOfRequiredParams = signature.params.filter(({ optional }) => !optional).length;
+    minNumberOfArgs = Math.min(minNumberOfArgs, numberOfRequiredParams);
+  }
+
+  const fieldMappingWithOneExtraArg = getFieldMapping(
+    signatureWithGreatestNumberOfParams.params
+  ).concat({
+    name: 'extraArg',
+    type: 'number',
+  });
+
+  // get the expected args from the first signature in case of errors
+  const hasOptionalArgs = minNumberOfArgs < maxNumberOfArgs;
+  const hasTooManyArgs = fieldMappingWithOneExtraArg.length > maxNumberOfArgs;
+
+  // the validation engine tries to be smart about signatures with optional args
+  let messageQuantifier = 'exactly ';
+  if (hasOptionalArgs && hasTooManyArgs) {
+    messageQuantifier = 'no more than ';
+  }
+  if (!hasOptionalArgs && !hasTooManyArgs) {
+    messageQuantifier = 'at least ';
+  }
+  testCases.set(
+    `from a_index | eval ${
+      getFunctionSignatures(
+        {
+          name,
+          ...defRest,
+          signatures: [
+            { ...signatureWithGreatestNumberOfParams, params: fieldMappingWithOneExtraArg },
+          ],
+        },
+        { withTypes: false }
+      )[0].declaration
+    }`,
+    [
+      `Error: [${name}] function expects ${messageQuantifier}${
+        maxNumberOfArgs === 1
+          ? 'one argument'
+          : maxNumberOfArgs === 0
+          ? '0 arguments'
+          : `${maxNumberOfArgs} arguments`
+      }, got ${fieldMappingWithOneExtraArg.length}.`,
+    ]
+  );
```

Comment:
Made more sense to do this once per function definition instead of once per signature.